crates/modelardb_compression/src/lib.rs

```diff
@@ -32,62 +32,155 @@ use modelardb_common::types::{
-    // as a TimestampArray and a ValueArray is the only supported input.
-    // However, as a result it is necessary to verify they have the same length.
+    // The uncompressed data must be passed as arrays instead of a RecordBatch as a TimestampArray
+    // and a ValueArray is the only supported input. However, as a result it is necessary to verify
+    // they have the same length.
     if uncompressed_timestamps.len() != uncompressed_values.len() {
         return Err(ModelarDbError::CompressionError(
             "Uncompressed timestamps and uncompressed values have different lengths.".to_owned(),
         ));
     }
 
-    // Enough memory for end_index elements are allocated to never require
-    // reallocation as one model is created per data point in the worst case.
+    // Enough memory for end_index compressed segments are allocated to never require reallocation
+    // as one compressed segments is created per data point in the absolute worst case.
     let end_index = uncompressed_timestamps.len();
     let mut compressed_record_batch_builder = CompressedSegmentBatchBuilder::new(end_index);
 
     // Compress the uncompressed timestamps and uncompressed values.
     let mut current_index = 0;
+    let mut residual_start_index = 0;
     while current_index < end_index {
-        // Create a compressed segment that represents the data points from
-        // current_index to index within error_bound where index <= end_index.
-        let compressed_segment_builder = CompressedSegmentBuilder::new(
-            univariate_id,
+        // Propose a model that represents the values from current_index to index within error_bound
+        // where index <= end_index.
+        let compressed_model_builder = CompressedModelBuilder::new(
             current_index,
             end_index,
             uncompressed_timestamps,
             uncompressed_values,
             error_bound,
         );
-        current_index = compressed_segment_builder.finish(&mut compressed_record_batch_builder);
+
+        let selected_model = compressed_model_builder.finish();
+
+        // The selected model is only stored as part of a compressed segment if it uses less storage
+        // space per value than the uncompressed values it represents.
+        if selected_model.bytes_per_value <= models::VALUE_SIZE_IN_BYTES as f32 {
+            // Compress residual data points with gorilla and store them in a compressed segments.
+            if residual_start_index != current_index {
+                compress_and_store_residual_value_range(
+                    univariate_id,
+                    residual_start_index,
+                    current_index - 1,
+                    uncompressed_timestamps,
+                    uncompressed_values,
+                    &mut compressed_record_batch_builder,
+                )
+            }
+
+            // Store the selected model and the corresponding timestamps in a compressed segments.
```

Comment:
Maybe `in a compressed segment.` instead.

Reply:
Fixed.